src/main/java/com/vmware/vipclient/i18n/util/LocaleUtility.java

```diff
@@ -147,43 +142,21 @@ public static boolean isSameLocale(String locale1, String locale2) {
     /*
      * pick up the matched locale from a locale list
      */
-    public static Locale pickupLocaleFromList(List<Locale> locales,
-            Locale preferredLocale) {
-        Locale langLocale = null;
-        preferredLocale = fmtToMappedLocale(preferredLocale);
-
-        // Use the first locale from the browser's list of preferred languages
-        // for the matching, so that it can keep the same way of getting locale
-        // with other VMware products, like vSphere Web Client, etc.
-        for (Locale configuredLocale : locales) {
-            // Language is matched
-            configuredLocale = fmtToMappedLocale(configuredLocale);
-            if (configuredLocale.getLanguage().equals(
-                    preferredLocale.getLanguage())) {
-                String configuredScript = configuredLocale.getScript();
-                String preferredScript = preferredLocale.getScript();
-                // Country is matched
-                if (((preferredScript.equalsIgnoreCase("")) && (configuredScript
-                        .equalsIgnoreCase("")))
-                        || ((!preferredScript.equalsIgnoreCase("")) && (preferredScript
-                                .equalsIgnoreCase(configuredScript)))) {
-                    return configuredLocale;
-                }
-                langLocale = langLocale == null ? configuredLocale : langLocale;
+    public static Locale pickupLocaleFromList(Set<Locale> locales,
+                                              Locale preferredLocale) {
+        Locale bestMatch = Locale.lookup(Arrays.asList(new Locale.LanguageRange(fmtToMappedLocale(preferredLocale).toLanguageTag())), locales);
+
+        // For any Chinese locale (zh-*) that is not supported (except for zh-Hans and zh-Hant), use the fallback locale even if "zh" is supported.
+        if (preferredLocale.getLanguage().equals("zh")) {
+            Locale zhLocale = fmtToMappedLocale(preferredLocale);
+            if (!locales.contains(zhLocale)) {
+                String langTag = zhLocale.toLanguageTag();
+                if (!langTag.equals("zh-Hans") && !langTag.equals("zh-Hant"))
+                    return null;
```

Comment:
```suggestion
		// With Chinese locale which is not configured/supported in web.xml, it
		// will return 'en_US' as default to meet the usage custom of Chinese,
		// e.g. for 'zh-HK' from client(browser) which is not
		// configured/supported yet, it will return 'en_US';
		// Other locale, like 'de-DE' 'ja-JP' etc.,
		// it will return 'de' 'ja'(main/parent language).
		if (bestMatch != null && bestMatch.getLanguage().equals("zh")) {
			if (!locales.contains(localeObject)) {
				return null;
			}
		}
```

Reply:
@Xiaochao8 , this change fails for test LocaleTest. testPickupLocaleFromListZh